File: 2023/day04/solution.py
#!/usr/bin/env python3
import sys

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case 'p1':
        PART1, PART2 = True, False
      case 'p2':
        PART1, PART2 = False, True
      case 'p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  cards = [
    {
      'win': {int(x)
              for x in parts[0].split()},
      'have': {int(x)
               for x in parts[1].split()},
    }
    for line in f.read().splitlines()
    if len(parts := line.split(': ')[1].split(' | ')) == 2
  ]
if DEBUG > 1: print(f'{cards=}')


def part1():
  total = 0
  for card in cards:
    intersection = card['win'].intersection(card['have'])
    value = (1 << len(intersection) - 1) if intersection else 0
    if DEBUG > 0: print(f'''win={card['win']} have={card['have']} {intersection=} {value=}''')
    total += value
  print(f'part 1: {total=}')


def part2():
  values = [len(x['win'].intersection(x['have'])) for x in cards]
  rows = len(cards)
  counts = [1] * len(values)
  if DEBUG > 1: print(f'{values=}')
  if DEBUG > 1: print(f'{counts=}')
  for i, value in enumerate(values.copy()):
    for j in range(value):
      if i + j + 1 >= rows: continue
      counts[i + j + 1] += counts[i]
      if DEBUG > 0: print(f'{i=} {value=} {j=} {counts[i + j+1]=}')
  total = sum(counts)
  if DEBUG > 1: print(f'{counts=}')
  print(f'part 2: {total=}')


if PART1: part1()
if PART2: part2()

File: 2023/day04/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

const parseInput = () => {
  const cards = input
    .replaceAll(' | ', ': ')
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => {
      const parts = line.split(/:\s+/);
      return {
        id: Number(parts[0].split(/\s+/)[1]),
        win: new Set(parts[1].split(/\s+/).map((token) => Number(token))),
        ours: new Set(parts[2].split(/\s+/).map((token) => Number(token))),
      };
    });
  return cards;
};

const part1 = () => {
  const cards = parseInput();
  const sum = cards.reduce((acc, card) => {
    const count = card.ours.intersection(card.win).size;
    const value = count ? 1 << (count - 1) : 0;
    debug(2, { card, count, value });
    return acc + value;
  }, 0);
  console.log('part 1:', sum);
};

const part2 = () => {
  const cards = parseInput();
  const multipliers: Map<number, number> = new Map(cards.map((card) => [card.id, 1]));
  debug(1, { multipliers });
  for (const card of cards) {
    const count = card.ours.intersection(card.win).size;
    const multiplier = multipliers.get(card.id)!;
    for (let cardId = card.id + 1; cardId < card.id + count + 1; ++cardId) {
      if (multipliers.has(cardId)) multipliers.set(cardId, multipliers.get(cardId)! + multiplier);
    }
  }
  debug(1, { multipliers });
  const sum = multipliers.values().reduce((acc, item) => acc + item, 0);
  console.log('part 2:', sum);
};

if (PART1) part1();
if (PART2) part2();
